clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3454,10 +3464,11 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc, SourceRange CallLoc,
   if (!LangOpts.SYCLIsDevice)
     return;
 
-  const CXXRecordDecl *KernelObj =
+  CXXRecordDecl *KernelObj =
```

Comment:
Is the loss of 'const' here inadvertent?

Reply:
Dropped it intentionally.
Looking at the documentation of  [const](https://en.cppreference.com/w/c/language/const)  and the fact that  [GetSYCLKernelObjectType](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L967) doesn't return a `const-qualified` type, I didn't see the need to have this `const-qualified`.
Besides, I see `const` qualifier [applied](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L3631) in some cases and [left](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L2511) other in others. 
If there is a rationale behind adding the `const` qualifier, I'd like to know.
If not, I can remove them at other places too.

Reply:
I think declaring something as `const` allows for compiler optimizations sometimes, but I'm not sure if those happen here. One advantage of declaring the KernelObj as const is it avoids us from accidentally modifying it. @premanandrao thoughts?

Reply:
I don't believe `const` can help optimizations: https://www.youtube.com/watch?v=FnGCDLhaxKU&t=5730s .

Reply:
Very interesting. Thanks for sharing @aelovikov-intel

Reply:
Not sure if this is of any relevance to why the KernelObject was `const` qualified, but the SYCL [spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interfaces.kernels.as.function-objects) mentions that : `The operator() member function must be const-qualified` , when kernels are defined as functors.

Reply:
> Dropped it intentionally. Looking at the documentation of [const](https://en.cppreference.com/w/c/language/const) and the fact that [GetSYCLKernelObjectType](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L967) doesn't return a `const-qualified` type, I didn't see the need to have this `const-qualified`. Besides, I see `const` qualifier [applied](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L3631) in some cases and [left](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L2511) out in other cases. If there is a rationale behind adding the `const` qualifier, I'd like to know. If not, I can remove them at other places too.

I would like the const re-applied in this case.  The rationale is for both visual affirmation that we are not modifying the object pointed to as well as for potential optimization.   (There are arguments that it does not really help with optimization; but it is not clear to me if that is a statement of the current state of the optimizer or a stronger statement that it can never be optimized.)

I would err on the other side.   I would apply const qualifiers to other places that do not have them but should; rather than remove it from all places.

Reply:
> I don't believe `const` can help optimizations: https://www.youtube.com/watch?v=FnGCDLhaxKU&t=5730s .

Thanks for that link @aelovikov-intel.   I remember seeing it a few years back, it is a good talk.  :-)

Reply:
> Not sure if this is of any relevance to why the KernelObject was `const` qualified, but the SYCL [spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interfaces.kernels.as.function-objects) mentions that : `The operator() member function must be const-qualified` , when kernels are defined as functors.

That wouldn't affect the internal compiler structure.   That const-ness is at the source level.